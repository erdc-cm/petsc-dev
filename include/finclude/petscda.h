!
!  $Id: petscda.h,v 1.20 2001/06/21 21:20:08 bsmith Exp $;
!
!  Include file for Fortran use of the DA (distributed array) package in PETSc
!
#if !defined (__PETSCDA_H)
#define __PETSCDA_H

#define DA PetscFortranAddr
#define DAPeriodicType integer
#define DAStencilType integer
#define DADirection integer
#define DMMG PetscFortranAddr
#define VecPack PetscFortranAddr
#define DM PetscFortranAddr

#define DALocalInfo        integer
!
!   DA_LOCAL_INFO_SIZE is one large than the size incase the DA is larger than an integer (on 64 bit systems)
!
#define DA_LOCAL_INFO_SIZE 22
#define DA_LOCAL_INFO_DIM 1
#define DA_LOCAL_INFO_DOF 2
#define DA_LOCAL_INFO_PT 4
#define DA_LOCAL_INFO_ST 5
#define DA_LOCAL_INFO_MX 6
#define DA_LOCAL_INFO_MY 7
#define DA_LOCAL_INFO_MZ 8
#define DA_LOCAL_INFO_XS 9
#define DA_LOCAL_INFO_YS 10
#define DA_LOCAL_INFO_ZS 11
#define DA_LOCAL_INFO_XM 12
#define DA_LOCAL_INFO_YM 13
#define DA_LOCAL_INFO_ZM 14
#define DA_LOCAL_INFO_GXS 15
#define DA_LOCAL_INFO_GYS 16
#define DA_LOCAL_INFO_GZS 17
#define DA_LOCAL_INFO_GXM 18
#define DA_LOCAL_INFO_GYM 19
#define DA_LOCAL_INFO_GZM 20
#define DA_LOCAL_INFO_DA  21

#define XG_RANGE in(DA_LOCAL_INFO_GXS)+1:in(DA_LOCAL_INFO_GXS)+in(DA_LOCAL_INFO_GXM)
#define YG_RANGE in(DA_LOCAL_INFO_GYS)+1:in(DA_LOCAL_INFO_GYS)+in(DA_LOCAL_INFO_GYM)
#define ZG_RANGE in(DA_LOCAL_INFO_GZS)+1:in(DA_LOCAL_INFO_GZS)+in(DA_LOCAL_INFO_GZM)
#define X_RANGE in(DA_LOCAL_INFO_XS)+1:in(DA_LOCAL_INFO_XS)+in(DA_LOCAL_INFO_XM)
#define Y_RANGE in(DA_LOCAL_INFO_YS)+1:in(DA_LOCAL_INFO_YS)+in(DA_LOCAL_INFO_YM)
#define Z_RANGE in(DA_LOCAL_INFO_ZS)+1:in(DA_LOCAL_INFO_ZS)+in(DA_LOCAL_INFO_ZM)

#define DAInterpolationType integer
#define DA_Q0 0
#define DA_Q1 1

#endif


#if !defined (PETSC_AVOID_DECLARATIONS)
!
!  Types of stencils
!
      integer DA_STENCIL_STAR,DA_STENCIL_BOX

      parameter (DA_STENCIL_STAR = 0,DA_STENCIL_BOX = 1)
!
!  Types of periodicity
!
      integer DA_NONPERIODIC,DA_XPERIODIC,DA_YPERIODIC,DA_XYPERIODIC
      integer DA_XYZPERIODIC,DA_XZPERIODIC,DA_YZPERIODIC,DA_ZPERIODIC

      parameter (DA_NONPERIODIC = 0,DA_XPERIODIC = 1,DA_YPERIODIC = 2)
      parameter (DA_XYPERIODIC = 3,DA_XYZPERIODIC = 4)
      parameter (DA_XZPERIODIC = 5,DA_YZPERIODIC = 6,DA_ZPERIODIC = 7)
!
! DA Directions
!
      integer DA_X,DA_Y,DA_Z

      parameter (DA_X = 0,DA_Y = 1,DA_Z = 2)
!
!  End of Fortran include file for the DA package in PETSc

#endif
