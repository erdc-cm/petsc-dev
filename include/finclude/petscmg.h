!
!  $Id: petscmg.h,v 1.11 2001/01/15 21:50:11 bsmith Exp $;
!
!  Include file for Fortran use of the MG preconditioner in PETSc
!
#if !defined (__PETSCMG_H)
#define __PETSCMG_H

#define MGType integer

#endif

#if !defined (PETSC_AVOID_DECLARATIONS)
!
!
      integer MGMULTIPLICATIVE,MGADDITIVE,MGFULL,MGKASKADE,MGCASCADE
      parameter (MGMULTIPLICATIVE=0,MGADDITIVE=1,MGFULL=2,MGKASKADE=3)
      parameter (MGCASCADE=3)

!
!  Other defines
!
      integer MG_V_CYCLE,MG_W_CYCLE
      parameter (MG_V_CYCLE=1,MG_W_CYCLE=2)

      external MGDEFAULTRESIDUAL
PETSC_DEC_ATTRIBUTES(MGDEFAULTRESIDUAL,'_MGDEFAULTRESIDUAL')

!
!     End of Fortran include file for the  MG include file in PETSc

#endif
