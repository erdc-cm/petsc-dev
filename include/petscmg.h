/* $Id: petscmg.h,v 1.21 2001/04/04 21:07:36 bsmith Exp $ */
/*
      Structure used for Multigrid preconditioners 
*/
#if !defined(__PETSCMG_H)
#define __PETSCMG_H
#include "petscsles.h"

/*E
    MGType - Determines the type of multigrid method that is run.

   Level: beginner

   Values:
+  MGMULTIPLICATIVE (default) - traditional V or W cycle as determined by MGSetCycles()
.  MGADDITIVE - the additive multigrid preconditioner where all levels are
                smoothed before updating the residual
.  MGFULL - same as multiplicative except one also performs grid sequencing, 
            that is starts on the coarsest grid, performs a cycle, interpolates
            to the next, performs a cycle etc
-  MGKASKADE - like full multigrid except one never goes back to a coarser level
               from a finer

.seealso: MGSetType()

E*/
typedef enum { MGMULTIPLICATIVE,MGADDITIVE,MGFULL,MGKASKADE } MGType;
#define MGCASCADE MGKASKADE;

#define MG_V_CYCLE     1
#define MG_W_CYCLE     2

EXTERN int MGSetType(PC,MGType);
EXTERN int MGCheck(PC);
EXTERN int MGSetLevels(PC,int,MPI_Comm*);
EXTERN int MGGetLevels(PC,int*);

EXTERN int MGSetNumberSmoothUp(PC,int);
EXTERN int MGSetNumberSmoothDown(PC,int);
EXTERN int MGSetCycles(PC,int);
EXTERN int MGSetCyclesOnLevel(PC,int,int);

EXTERN int MGGetSmoother(PC,int,SLES*);
EXTERN int MGGetSmootherDown(PC,int,SLES*);
EXTERN int MGGetSmootherUp(PC,int,SLES*);
EXTERN int MGGetCoarseSolve(PC,SLES*);

EXTERN int MGSetRhs(PC,int,Vec);
EXTERN int MGSetX(PC,int,Vec);
EXTERN int MGSetR(PC,int,Vec);

EXTERN int MGSetRestriction(PC,int,Mat);
EXTERN int MGSetInterpolate(PC,int,Mat);
EXTERN int MGSetResidual(PC,int,int (*)(Mat,Vec,Vec,Vec),Mat);
EXTERN int MGDefaultResidual(Mat,Vec,Vec,Vec);


#endif

