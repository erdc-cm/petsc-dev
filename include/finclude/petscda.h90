!
!
!  Additional DA include file for use of PETSc with Fortran 90/HPF
!
#if !defined (PETSC_AVOID_DECLARATIONS)
#include "finclude/ftn-custom/petscda.h90"
#endif

