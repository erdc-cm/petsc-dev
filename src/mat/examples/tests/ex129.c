
/*
  Laplacian in 3D. Use for testing MatSolve routines. 
  Modeled by the partial differential equation

   - Laplacian u = 1,0 < x,y,z < 1,

   with boundary conditions
   u = 1 for x = 0, x = 1, y = 0, y = 1, z = 0, z = 1.
*/

static char help[] = "This example is for testing different MatSolve routines,namely MatSolve,MatSolveAdd,MatSolveTranspose and MatSolveTransposeAdd.\n\
Example usage: ./ex129 -solve_type 1 -dof 1\n\n";

#include "petscda.h"
#include "petscmg.h"

extern PetscErrorCode ComputeMatrix(DA,Mat);
extern PetscErrorCode ComputeRHS(DA,Vec);

#undef __FUNCT__
#define __FUNCT__ "main"
int main(int argc,char **argv)
{
  PetscErrorCode    ierr;
  PetscMPIInt       size;
  Vec               x,b,y,b1;
  DA                da;
  Mat               A,F;
  MatFactorInfo     info;
  IS                perm,iperm;
  PetscInt          dof=1,M=-8,solve_type=0;
  PetscScalar       one = 1.0;
  PetscReal         norm;

  PetscInitialize(&argc,&argv,(char *)0,help);
  ierr = MPI_Comm_size(PETSC_COMM_WORLD,&size);CHKERRQ(ierr);
  if(size != 1) SETERRQ(1,"This is a uniprocessor example only\n");
  ierr = PetscOptionsGetInt(PETSC_NULL,"-dof",&dof,PETSC_NULL);CHKERRQ(ierr);
  ierr = PetscOptionsGetInt(PETSC_NULL,"-M",&M,PETSC_NULL);CHKERRQ(ierr);

  ierr = DACreate(PETSC_COMM_WORLD,&da);CHKERRQ(ierr);
  ierr = DASetDim(da,3);CHKERRQ(ierr);
  ierr = DASetPeriodicity(da,DA_NONPERIODIC);CHKERRQ(ierr);
  ierr = DASetStencilType(da,DA_STENCIL_STAR);CHKERRQ(ierr);
  ierr = DASetSizes(da,M,M,M);CHKERRQ(ierr);
  ierr = DASetNumProcs(da,PETSC_DECIDE,PETSC_DECIDE,PETSC_DECIDE);CHKERRQ(ierr);
  ierr = DASetDof(da,dof);CHKERRQ(ierr);
  ierr = DASetStencilWidth(da,1);CHKERRQ(ierr);
  ierr = DASetVertexDivision(da,PETSC_NULL,PETSC_NULL,PETSC_NULL);CHKERRQ(ierr);
  ierr = DASetFromOptions(da);CHKERRQ(ierr);

  ierr = DACreateGlobalVector(da,&x);CHKERRQ(ierr);
  ierr = DACreateGlobalVector(da,&b);CHKERRQ(ierr);
  ierr = VecDuplicate(b,&y);CHKERRQ(ierr);
  ierr = ComputeRHS(da,b);CHKERRQ(ierr);
  ierr = VecSet(y,one);CHKERRQ(ierr);
  ierr = DAGetMatrix(da,MATBAIJ,&A);CHKERRQ(ierr);
  ierr = ComputeMatrix(da,A);CHKERRQ(ierr);

  ierr = MatGetOrdering(A,MATORDERING_ND,&perm,&iperm);CHKERRQ(ierr);
  ierr = MatGetFactor(A,MAT_SOLVER_PETSC,MAT_FACTOR_LU,&F);CHKERRQ(ierr);
  info.fill = 5.0;
  ierr = MatLUFactorSymbolic(F,A,perm,iperm,&info);CHKERRQ(ierr);
  ierr = MatLUFactorNumeric(F,A,&info);CHKERRQ(ierr);
  ierr = PetscOptionsGetInt(PETSC_NULL,"-solve_type",&solve_type,PETSC_NULL);CHKERRQ(ierr);
  ierr = VecDuplicate(y,&b1);CHKERRQ(ierr);
  switch (solve_type){
  case 1:
    /* MatSolve */
    ierr = MatSolve(F,b,x);CHKERRQ(ierr);
    ierr = MatMult(A,x,b1);CHKERRQ(ierr);
    ierr = VecAXPY(b1,-1.0,b);CHKERRQ(ierr);
    ierr = VecNorm(b1,NORM_2,&norm);CHKERRQ(ierr);
    ierr = PetscPrintf(PETSC_COMM_WORLD,"Error of norm %g\n",norm);CHKERRQ(ierr);
    break;
  case 2:
    /* MatSolveTranspose */
    ierr = MatSolveTranspose(F,b,x);CHKERRQ(ierr);
    ierr = MatMultTranspose(A,x,b1);CHKERRQ(ierr);
    ierr = VecAXPY(b1,-1.0,b);CHKERRQ(ierr);
    ierr = VecNorm(b1,NORM_2,&norm);CHKERRQ(ierr);
    ierr = PetscPrintf(PETSC_COMM_WORLD,"Error of norm %g\n",norm);CHKERRQ(ierr);
    break;
  case 3:
    /* MatSolveAdd */
    ierr = MatSolveAdd(F,b,y,x);CHKERRQ(ierr);
    ierr = MatMult(A,y,b1);CHKERRQ(ierr);
    ierr = VecScale(b1,-1.0);CHKERRQ(ierr);
    ierr = MatMultAdd(A,x,b1,b1);
    ierr = VecAXPY(b1,-1.0,b);CHKERRQ(ierr);
    ierr = VecNorm(b1,NORM_2,&norm);CHKERRQ(ierr);
    ierr = PetscPrintf(PETSC_COMM_WORLD,"Error or norm %g\n",norm);CHKERRQ(ierr);
    break;
  case 4:
    /* MatSolveTransposeAdd */
    ierr = MatSolveTransposeAdd(F,b,y,x);CHKERRQ(ierr); 
    ierr = MatMultTranspose(A,y,b1);CHKERRQ(ierr);
    ierr = VecScale(b1,-1.0);CHKERRQ(ierr);
    ierr = MatMultTransposeAdd(A,x,b1,b1);
    ierr = VecAXPY(b1,-1.0,b);CHKERRQ(ierr);
    ierr = VecNorm(b1,NORM_2,&norm);CHKERRQ(ierr);
    ierr = PetscPrintf(PETSC_COMM_WORLD,"Error or norm %g\n",norm);CHKERRQ(ierr);
    break;
  }
  ierr = VecDestroy(x);CHKERRQ(ierr);
  ierr = VecDestroy(b);CHKERRQ(ierr);
  ierr = VecDestroy(b1);CHKERRQ(ierr);
  ierr = VecDestroy(y);CHKERRQ(ierr);
  ierr = MatDestroy(A);CHKERRQ(ierr);
  ierr = MatDestroy(F);CHKERRQ(ierr);
  ierr = ISDestroy(perm);CHKERRQ(ierr);
  ierr = ISDestroy(iperm);CHKERRQ(ierr);
  ierr = DADestroy(da);CHKERRQ(ierr);
  ierr = PetscFinalize();CHKERRQ(ierr);
  return 0;
}

#undef __FUNCT__
#define __FUNCT__ "ComputeRHS"
PetscErrorCode ComputeRHS(DA da,Vec b)
{
  PetscErrorCode ierr;
  PetscInt       mx,my,mz;
  PetscScalar    h;

  PetscFunctionBegin;
  ierr = DAGetInfo(da,0,&mx,&my,&mz,0,0,0,0,0,0,0);CHKERRQ(ierr);
  h    = 1.0/((mx-1)*(my-1)*(mz-1));
  ierr = VecSet(b,h);CHKERRQ(ierr);
  PetscFunctionReturn(0);
}
    
#undef __FUNCT__
#define __FUNCT__ "ComputeMatrix"
PetscErrorCode ComputeMatrix(DA da,Mat B)
{
  PetscErrorCode ierr;
  PetscInt       i,j,k,mx,my,mz,xm,ym,zm,xs,ys,zs,dof,k1,k2,k3;
  PetscScalar    *v,*v_neighbor,Hx,Hy,Hz,HxHydHz,HyHzdHx,HxHzdHy,r1,r2;
  MatStencil     row,col;
  PetscRandom    rand; 

  ierr = PetscRandomCreate(PETSC_COMM_WORLD,&rand);CHKERRQ(ierr);
  ierr = PetscRandomSetType(rand,PETSCRAND);CHKERRQ(ierr);
  ierr = PetscRandomSetSeed(rand,1);CHKERRQ(ierr);
  ierr = PetscRandomSetInterval(rand,-.001,.001);CHKERRQ(ierr);
  ierr = PetscRandomSetFromOptions(rand);CHKERRQ(ierr);

  ierr = DAGetInfo(da,0,&mx,&my,&mz,0,0,0,&dof,0,0,0);CHKERRQ(ierr); 
  /* For simplicity, this example only works on mx=my=mz */
  if ( mx != my || mx != mz) SETERRQ3(1,"This example only works with mx %d = my %d = mz %d\n",mx,my,mz);

  Hx = 1.0 / (PetscReal)(mx-1); Hy = 1.0 / (PetscReal)(my-1); Hz = 1.0 / (PetscReal)(mz-1);
  HxHydHz = Hx*Hy/Hz; HxHzdHy = Hx*Hz/Hy; HyHzdHx = Hy*Hz/Hx;

  ierr = PetscMalloc((2*dof*dof+1)*sizeof(PetscScalar),&v);CHKERRQ(ierr);
  v_neighbor = v + dof*dof;
  ierr = PetscMemzero(v,(2*dof*dof+1)*sizeof(PetscScalar));CHKERRQ(ierr);
  k3 = 0;
  for (k1=0; k1<dof; k1++){
    for (k2=0; k2<dof; k2++){
      if (k1 == k2){
        v[k3]          = 2.0*(HxHydHz + HxHzdHy + HyHzdHx);
        v_neighbor[k3] = -HxHydHz;
      } else {
	ierr = PetscRandomGetValue(rand,&r1);CHKERRQ(ierr);
	ierr = PetscRandomGetValue(rand,&r2);CHKERRQ(ierr);
	v[k3] = r1;
	v_neighbor[k3] = r2;
      }	
      k3++;
    }
  }
  ierr = DAGetCorners(da,&xs,&ys,&zs,&xm,&ym,&zm);CHKERRQ(ierr);
  
  for (k=zs; k<zs+zm; k++){
    for (j=ys; j<ys+ym; j++){
      for(i=xs; i<xs+xm; i++){
        row.i = i; row.j = j; row.k = k;
	if (i==0 || j==0 || k==0 || i==mx-1 || j==my-1 || k==mz-1){ /* boudary points */	 
	  ierr = MatSetValuesBlockedStencil(B,1,&row,1,&row,v,INSERT_VALUES);CHKERRQ(ierr);
        } else { /* interior points */
          /* center */
          col.i = i; col.j = j; col.k = k;
          ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v,INSERT_VALUES);CHKERRQ(ierr);          
          
          /* x neighbors */
	  col.i = i-1; col.j = j; col.k = k;
          ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v_neighbor,INSERT_VALUES);CHKERRQ(ierr);
	  col.i = i+1; col.j = j; col.k = k;
	  ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v_neighbor,INSERT_VALUES);CHKERRQ(ierr);
	 
	  /* y neighbors */
	  col.i = i; col.j = j-1; col.k = k;
	  ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v_neighbor,INSERT_VALUES);CHKERRQ(ierr);
	  col.i = i; col.j = j+1; col.k = k;
	  ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v_neighbor,INSERT_VALUES);CHKERRQ(ierr);
	 
          /* z neighbors */
	  col.i = i; col.j = j; col.k = k-1;
	  ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v_neighbor,INSERT_VALUES);CHKERRQ(ierr);
	  col.i = i; col.j = j; col.k = k+1;
          ierr = MatSetValuesBlockedStencil(B,1,&row,1,&col,v_neighbor,INSERT_VALUES);CHKERRQ(ierr);
        }
      }
    }
  }
  ierr = MatAssemblyBegin(B,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
  ierr = MatAssemblyEnd(B,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
  ierr = PetscFree(v);CHKERRQ(ierr);
  ierr = PetscRandomDestroy(rand);CHKERRQ(ierr);
  /* ierr = MatView(B,PETSC_VIEWER_STDOUT_WORLD);CHKERRQ(ierr); */ 
  return 0;
}

