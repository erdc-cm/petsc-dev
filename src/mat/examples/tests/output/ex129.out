MatSolve:             Error of norm < 1.e-12
MatSolveTranspose :   Error of norm < 1.e-12
MatSolveAdd:          Error or norm < 1.e-12
MatSolveTransposeAdd: Error or norm < 1.e-12
